miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/SampleRestController.java

```diff
@@ -12,64 +12,287 @@
+      Long topProcessingClassId = dto.getSampleClassId();
+      if (sampleDto instanceof SampleLCMTubeDto) {
+        SampleLCMTubeDto lcmDto = (SampleLCMTubeDto) dto;
+        // Some hierarchies have two Tissue Processing levels
+        lcmDto.setParentSlideClassId(inferIntermediateSampleClassId(dto, dto.getSampleClassId(), SampleTissueProcessing.CATEGORY_NAME,
+            SampleTissueProcessing.CATEGORY_NAME, true));
+        if (lcmDto.getParentSlideClassId() != null) {
+          topProcessingClassId = lcmDto.getParentSlideClassId();
+        }
+      }
+      dto.setParentTissueSampleClassId(
+          inferIntermediateSampleClassId(dto, topProcessingClassId, SampleTissueProcessing.CATEGORY_NAME,
+              SampleTissue.CATEGORY_NAME, false));
+    }
+    return Dtos.to(sampleDto);
+  }
+
+  private Long inferIntermediateSampleClassId(DetailedSampleDto dto, Long childClassId,
+      String childClassCategory, String parentCategory, boolean nullOk) {
+    if (dto.getParentId() != null) {
+      return null;
+    }
+    if (childClassId == null) {
+      throw new RestException("No parent and no target sample class.", Status.BAD_REQUEST);
+    }
+    // infer parent class
+    SampleClass parentClass = sampleClassService.inferParentFromChild(childClassId, childClassCategory, parentCategory);
+    if (parentClass == null && !nullOk) {
+      throw new IllegalStateException(String.format("%s class with id %d has no %s parents", childClassCategory, childClassId,
+          parentCategory));
+    }
+    return parentClass == null ? null : parentClass.getId();
   }
 
-  @RequestMapping(value = "{sampleId}", method = RequestMethod.GET, produces="application/json")
-  public @ResponseBody String getSampleById(@PathVariable Long sampleId) throws IOException {
-    ObjectMapper mapper = new ObjectMapper();
-    Sample s = sampleService.get(sampleId);
-    if (s == null) {
-      throw new RestException("No sample found with ID: " + sampleId, Status.NOT_FOUND);
+  @RequestMapping(value = "/{id}", method = RequestMethod.PUT, headers = { "Content-type=application/json" })
+  @ResponseBody
+  @ResponseStatus(HttpStatus.OK)
+  public SampleDto updateSample(@PathVariable("id") Long id, @RequestBody SampleDto sampleDto, UriComponentsBuilder b) throws IOException {
+    if (sampleDto == null) {
+      log.error("Received null sampleDto from front end; cannot convert to Sample. Something likely went wrong in the JS DTO conversion.");
+      throw new RestException("Cannot convert null to Sample", Status.BAD_REQUEST);
     }
-    return mapper.writeValueAsString(s);
+    Sample sample = Dtos.to(sampleDto);
+    sample.setId(id);
+    sampleService.update(sample);
+    return getSample(id, b);
+  }
+
+  @RequestMapping(value = "/{id}", method = RequestMethod.DELETE)
+  @ResponseBody
+  public ResponseEntity<?> deleteSample(@PathVariable("id") Long id, HttpServletResponse response) throws IOException {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Remove usage of generic wildcard type. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1452)